### PR TITLE
NIFI-6400 Better options, consistent ids for ShellUserGroupProvider.

### PR DESCRIPTION
**Pull request by natural**
#### Description of PR

The code in this change-set provides the following:

- consistent user and group identifiers from named UUIDs
- two additional regular expression properties for excluding groups and users
- new and updated integration tests for the above

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 1]
```diff
@@ -255,11 +256,19 @@ public void onConfigured(AuthorizerConfigurationContext configurationContext) th
...
+        } catch (final PatternSyntaxException e) {
+            throw new AuthorizerCreationException(e);
+        }
+
         // With our command set selected, and our system check passed, we can pull in the users and groups:
         refreshUsersAndGroups();
 
         // And finally, our last init step is to fire off the refresh thread:
-        scheduler.scheduleWithFixedDelay(this::refreshUsersAndGroups, initialDelay, fixedDelay, TimeUnit.SECONDS);
+        scheduler.scheduleWithFixedDelay(this::refreshUsersAndGroups, fixedDelay, fixedDelay, TimeUnit.SECONDS);
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: `refreshUsersAndGroups` cannot allow any exceptions to escape. If one does, the scheduling will not continue. We typically have a `try {...} catch (Throwable t) {...}` to ensure this does not happen.

Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: For instance `LdapUserGroupProvider` [1]

[1] https://github.com/apache/nifi/blob/2491c51d34a4012dd6c9058f60f80ea2203e311b/nifi-nar-bundles/nifi-ldap-iaa-providers-bundle/nifi-ldap-iaa-providers/src/main/java/org/apache/nifi/ldap/tenants/LdapUserGroupProvider.java#L395

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 2]
```diff
@@ -525,26 +549,29 @@ private void reconcilePrimaryGroups(Map<String, User> uidToUser, Map<String, Gro
...
                 logger.warn("user: " + primaryUser + " primary group not found");
-            } else {
+            } else if (!excludeGroups.matcher(primaryGroup.getName()).matches()) {
                 Set<String> groupUsers = primaryGroup.getUsers();
                 if (!groupUsers.contains(primaryUser.getIdentity())) {
                     Set<String> secondSet = new HashSet<>(groupUsers);
                     secondSet.add(primaryUser.getIdentity());
-                    Group group = new Group.Builder().name(primaryGroup.getName()).identifier(primaryGid).addUsers(secondSet).build();
-                    gidToGroup.put(primaryGid, group);
+                    String groupIdentifier = getNameBasedUUID(primaryGid);
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: Can we use `Group.Builder.identifierGenerateFromSeed`?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 3]
```diff
@@ -499,12 +522,13 @@ private void rebuildGroups(List<String> groupLines, Map<String, Group> groupsByI
                     logger.error("list membership shell exception: " + ioexc);
                 }
 
-                if (name != null && id != null && !name.equals("") && !id.equals("")) {
-                    Group group = new Group.Builder().name(name).identifier(id).addUsers(users).build();
-                    groupsById.put(id, group);
+                if (name != null && id != null && !name.equals("") && !id.equals("") && !excludeGroups.matcher(name).matches()) {
+                    String groupIdentifier = getNameBasedUUID(id);
+                    Group group = new Group.Builder().name(name).identifier(groupIdentifier).addUsers(users).build();
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: Can we use `Group.Builder.identifierGenerateFromSeed`?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 4]
```diff
@@ -445,20 +467,21 @@ private void rebuildUsers(List<String> userLines, Map<String, User> idToUser, Ma
             if (record.length > 2) {
                 String name = record[0], id = record[1], gid = record[2];
 
-                if (name != null && id != null && !name.equals("") && !id.equals("")) {
-
-                    User user = new User.Builder().identity(name).identifier(id).build();
-                    idToUser.put(id, user);
+                if (name != null && id != null && !name.equals("") && !id.equals("") && !excludeUsers.matcher(name).matches()) {
+                    String identifier = getNameBasedUUID(id);
+                    User user = new User.Builder().identity(name).identifier(identifier).build();
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: Can we use `User.Builder.identifierGenerateFromSeed`?

**Comment by mcgilman** [comment]
@natural Thanks for the update! We should be able to leverage the methods in `User.Builder` and `Group.Builder` to remove the private method for `getNameBasedUUID`. There is admittedly one scenario where the `Group` that is being created is only needed temporarily but I think this would use the APIs for as the were expected. I've typed up an example here [1]. Happy to discuss it and work through it more.

[1] https://github.com/mcgilman/nifi/commit/c423a016cc00816436191962a4eda9445f7abe17

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/authorizers.xml [hunk 5]
```diff
@@ -171,14 +171,15 @@
...
         'Refresh Delay' - duration to wait between subsequent refreshes.  Default is '5 mins'.
+        'Exclude Groups' - regular expression used to exclude groups.  Default is '', which means no groups are excluded.
+        'Exclude Users' - regular expression used to exclude users.  Default is '', which means no users are excluded.
     -->
     <!-- To enable the shell-user-group-provider remove 2 lines. This is 1 of 2.
     <userGroupProvider>
         <identifier>shell-user-group-provider</identifier>
         <class>org.apache.nifi.authorization.ShellUserGroupProvider</class>
-        <property name="Initial Refresh Delay">5 mins</property>
+        <property name="Exclude Users">^s.*</property>
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/authorizers.xml]: I am +1 on the empty string `''` as a default value.

**Comment by mcgilman** [comment]
Thanks @natural! This has been merged to master.

**Pull request closed without merging**